### DISCUSSION TITLE
PyScript REPL + WASM Integration

### DISCUSSION THREAD

**Discussion by nyckmaia** [pyscript/pyscript#1180] [Proposals]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other feature requests and couldn't find a duplicate (including also the ``type-feature`` tag)
- [X] I confirmed that it's not related to another project are area (see the above section)

### What is the idea?

Would be great if some how there is a integration between PyScript REPL and WebAssembly.
So, the PyScript REPL user will be able to call Python and JS/WebAssembly libraries functions/objects.

### Why is this needed

There are a lot of custom C/C++ libraries that would be great to use with PyScript.
I know that Pyodide is doing this interface between PyScript and some big C/C++ libraries (as NumPy and Matplotlib), but I would like to be able to compile my own C/C++ libraries to WebAssembly and be able to call the functions and objects in a REPL with PyScript.

I know that a pure JS REPL already exists, but the WebAssembly  library functions/objects cannot be mix with the Python code.
Example:
My WebAssembly library computes a large dataset and I would like to plot it using Matplotlib in a PyScript REPL.

### What should happen?

PyScript REPL should be able to run both: Python and WebAssembly code

**Comment by rth** [comment]
> but the WebAssembly library functions/objects cannot be mix with the Python code.

Could you provide more information about how you are mixing them and what kind of errors you are getting when it fails? If your WASM library exposes a JS API (or returns JS data) you should be able to call it from Python via the Pyodide FFI.  Otherwise, how would you call a C/C++ library from Python? If it's via ctypes or cffi, then it's likely indeed that they need to have binary compatibility in terms of how they were built.

**Comment by nyckmaia** [comment]
Okay, maybe I'm not quite understanding how to get this interaction between the PyScript REPL and the WebAssembly modules.

To facilitate the conversation, I'll show you a practical example:
I compiled a `C++` class called `Dog` for WebAssembly, generating the file `dog.wasm`.

From the official PyScript REPL example, I added a `dog.wasm` file to the `index.html` file.

To test the import of this WASM module, I added buttons that invoke 2 public methods of this `Dog` class.

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width,initial-scale=1" />

    <title>Custom REPL Example</title>

    <link rel="icon" type="image/png" href="favicon.png" />
    <link rel="stylesheet" href="https://pyscript.net/alpha/pyscript.css" />
    <link rel="stylesheet" href="repl.css" />

    <script defer src="https://pyscript.net/alpha/pyscript.js"></script>
  </head>

  <body>
    <button id="hi">Say hi</button>
    <button id="bark">Bark (console)</button>
      <h1 class="font-semibold text-2xl ml-5">Custom REPL</h1>
      <py-box widths="1/2;1/2">
        <py-repl id="my-repl" auto-generate="true" std-out="output" std-err="err-div"> </py-repl>
        <div id="output" class="p-4"></div>
      </py-box>
      <footer id="err-div" class="bg-red-700 text-white text-center border-t-4 border-green-500 fixed inset-x-0 bottom-0 p-4 hidden">
    </footer>
  </body>

  <script src="/dog.wasm"></script>
  <script>
    Module.addOnPostRun(() => {
      const dog = new Module.Dog('Snoopy')
      hi.onclick = () => alert(dog.getGreeting())
      bark.onclick = () => dog.makeSound()
    })
  </script>
</html>
```

I would like to invoke the WAMS module via the PyScript REPL, so that it would be possible to instantiate an object of class `Dog` via the RELP.

I imagine it would be something like:
```
myDog = Module.Dog("Foo")
myDog.makeSound()
```
**Question: Is there any way to use PyScript REPL to invoke modules/objects/functions from a WebAssembly file as shown above?**

**Comment by rth** [comment]
Assuming that Module is a global JS variable, could you try something like the following in the REPL?
```python
from js import Module

myDog = Module.Dog.new("Foo")
myDog.makeSound()
```

For more detail, see [Pyodide type conversions docs](https://pyodide.org/en/stable/usage/type-conversions.html#proxying-from-javascript-into-python).

**Comment by nyckmaia** [comment]
Great! It works!

Thank you very much!

**Comment by nyckmaia** [comment]
@rth  Considering the simple example of the `Dog` class cited above, and considering that I have successfully compiled it for `WASM`, what is needed for me to create an instance/object of the `Dog` class using Python's native syntax right in the `PyScript REPL`?

Example:
```
# PyScript REPL

myDog = Dog("Foo")  # without the 'new' keyword
```

Today, to use my `Dog` class in the `PyScript REPL` I have to use this syntax below:
```
# PyScript REPL

from js import Module

myDog = Module.Dog.new("Foo")  # Must have this 'new' keyword
```
**Question:** What I have to do to be able to use the native Python syntax like the first example above?

Thank you,

**Comment by marimeireles** [comment]
@antocuni might be interested in seeing this issue.

**Comment by antocuni** [comment]
@nyckmaia note that there is no different between the PyScripy REPL and `<py-script>` blocks in terms of capabilities: if you can do it inside a `<py-script>`, you can do it also inside the REPL.

For the future, I agree that it would be nice to have a more direct  py <=> WASM interaction. I never thought of it but I think it's actually a good idea eventually, because it would allow a more efficient wrapping compared to passing through JS. Thanks for bringing it up!

However, it's not straightforward, for many reasons: one is that wrapping complex a C API is hard: it seems simple with simple examples, but e.g. if you return a pointer to a struct, you need a way to read/write the fields of that struct, and for this you need something like `cffi`. It is probably possible to integrate cffi with WASM but it's a multi week project, for sure.

I think that for now the best way to call WASM from Python is by using the `js` module as suggested by Roman.
If you really need a more Pythonic way of calling it, I fear that the best solution is to write a wrapper by hand:
```python
import js
def Dog(name):
    return js.Module.Dog.new(name)
```

Btw, OT question: how did you compile your C++ file to WASM? It seems that the toolchain that you used automatically creates a JS wrapper for it: did you use `embind` or something else?

**Comment by nyckmaia** [comment]
Thank you @antocuni and @marimeireles,

Responses
------------

>  I agree that it would be nice to have a more direct py <=> WASM interaction.

Yes, if is possible, the interaction between `PyScript` and `WASM` should be a great 'game change' in my view.
Please, let me know the news about it in the future, ok?

>  I fear that the best solution is to write a wrapper by hand:

Thank you for this suggestion. I will think about it.

> how did you compile your C++ file to WASM?

Yes, In this simple example above I used the `embind` to get the `dog.wasm` and the `dog.js` wrapper file.
You can see the `embind` C++ class wrapper tutorial [here](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#classes).

Question
----------

As you said, I understand the complexity to enable a easily `PyScript <=> WASM` communication.
In another hand, today we can use `C` and `C++` libraries inside `PyScript` without change the `API`.
An example of `C++` library in Pyodide is `boost-cpp`. Link [here](https://github.com/pyodide/pyodide/tree/main/packages/boost-cpp).

If I understand well, the main ideia of running `C` and `C++` on web browser is the possibility to compile all to `WASM`, as you can see in this picture below:

![image](https://user-images.githubusercontent.com/4555404/179739717-42263502-a604-4097-9205-3b69f8237dbb.png)

So, I don't understand what is the difference between my build process and `Pyodide` build process.
When I build, I got a API change, that I need to use the `.new(name)` that I don't like it.

I tried to build my own `Dog` C++ class with Pyodide, but I don't understand it.

Finally
--------
So, what is missing?
How can `Pyodide` packages be used with `PyScript` without any API change?
How can I do the same build process to my simple `Dog` class?

Thank you

**Comment by antocuni** [comment]
> So, what is missing?

what is missing is the equivalent of `embind` for Python.
You are not just compiling C++ code: with `embind`, you compile C++ code *PLUS* automatically generated bindings for JS.
What you would like is native python bindings for your C++ code: there are many projects to do that, e.g. `pybind11`, `nanobind`, `cppyy` or `cffi` for pure C (not C++). But probably none of them work out of the box in the context of WASM, so there is probably work to be done in order to make them usable.

This is certainly inside the scope of PyScript, but I don't think it's our top priority for now, especially because you can easily work around the limitation using `embind` + `import js`.

**Comment by nyckmaia** [comment]
Ok! Now I got it!

Topic 01
---------
At this time, I use `Pybind11` to use my `C++` libraries inside the `Python` environment. It works great!
In the future, would be great if we can have a build process that integrates `C++` + `Pybind11` and `PyScript`. I think that will solve everything!

Topic 02
---------
Do you know if is this possible to do using `Pyodide`? Like the `boost-cpp` example above?
In another words: Is there a way to use `Pyodide` in my `Dog` class to enable use it inside `PyScript` without changed API `.new(name)`?
If yes, how can I do that with my simple `Dog` class?

**Comment by antocuni** [comment]
> Do you know if is this possible to do using Pyodide? Like the boost-cpp example above?

It might be possible if you build your module together with pyodide, as described here:
https://pyodide.org/en/stable/development/new-packages.html

but this is more a question for pyodide, not pyscript. I guess @rth can give you a more detailed answer.
